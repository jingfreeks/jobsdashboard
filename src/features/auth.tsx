import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from '../config/store';

interface AuthState {
  user: string | null;
  token: string | null;
  userId: string | null;
  roles: string[];
  onboardingComplete: boolean;
}

interface CredentialsPayload {
  user: string;
  accessToken: string;
  userId: string;
  roles: string[];
}

const initialState: AuthState = {
  user: null,
  token: null,
  userId: null,
  roles: [],
  onboardingComplete: false,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    setCredentials: (state, action: PayloadAction<CredentialsPayload>) => {
      const { user, accessToken, userId, roles } = action.payload;
      
      // Validate the payload
      if (!user || !accessToken || !userId) {
        console.warn('Invalid credentials payload:', action.payload);
        return;
      }
      
      state.user = user;
      state.token = accessToken;
      state.userId = userId;
      state.roles = Array.isArray(roles) ? roles : [];
    },
    setLogout: (state) => {
      state.user = null;
      state.token = null;
      state.userId = null;
      state.roles = [];
      state.onboardingComplete = false;
    },
    // Action to handle rehydration from storage
    rehydrateAuth: (state, action: PayloadAction<Partial<AuthState>>) => {
      const { user, token, userId, roles, onboardingComplete } = action.payload;
      if (user) state.user = user;
      if (token) state.token = token;
      if (userId) state.userId = userId;
      if (roles) state.roles = Array.isArray(roles) ? roles : [];
      if (onboardingComplete !== undefined) state.onboardingComplete = onboardingComplete;
    },
    // Action to set onboarding completion
    setOnboardingComplete: (state, action: PayloadAction<boolean>) => {
      state.onboardingComplete = action.payload;
    },
  },
});

export const { setCredentials, setLogout, rehydrateAuth, setOnboardingComplete } = authSlice.actions;

export default authSlice.reducer;

export const selectCurrentUser = (state: RootState) => state.auth.user;
export const selectCurrentUserId = (state: RootState) => state.auth.userId;
export const selectCurrentToken = (state: RootState) => state.auth.token;
export const selectUserRoles = (state: RootState) => state.auth.roles;
export const selectIsAuthenticated = (state: RootState) => !!state.auth.token;
export const selectOnboardingComplete = (state: RootState) => state.auth.onboardingComplete;