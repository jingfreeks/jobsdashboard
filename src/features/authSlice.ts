import {apiSlice} from '@/config/apiSplice';

export const authApiSlice = apiSlice.injectEndpoints({
  endpoints: builder => ({
    login: builder.mutation({
      query: credentials => ({
        url: '/auth',
        method: 'POST',
        body: {...credentials},
      }),
    }),
    logout: builder.mutation({
      query: () => ({
        url: '/logout',
        method: 'POST',
        // body: {...credentials},
      }),
    }),
    signup: builder.mutation({
      query: credentials => ({
        url: '/signup',
        method: 'POST',
        body: {...credentials},
      }),
    }),
  }),
  overrideExisting: true,
});

export const {useLoginMutation,useSignupMutation,useLogoutMutation} = authApiSlice || {};