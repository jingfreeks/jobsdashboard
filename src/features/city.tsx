import { apiSlice } from '@/config/apiSplice';

// Type definitions
export interface City {
  _id: string;
  name: string;
  stateId?: string;
  statename?: string;
  image?: string;
}

export interface CityFormData {
  name: string;
  stateId?: string;
  image?: string;
}

export interface UpdateCityData {
  _id: string;
  name: string;
  stateId?: string;
  image?: File;
}

export const cityApiSlice = apiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getCities: builder.query<City[], void>({
      query: () => '/city',
      transformResponse: (response: City[]) => {
        // Sort cities alphabetically by name
        return response.sort((a, b) => a.name.localeCompare(b.name));
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['City'] as any,
      keepUnusedDataFor: 300, // Cache for 5 minutes
    }),
    addCity: builder.mutation<City, CityFormData>({
      query: (credentials) => ({
        url: '/city',
        method: 'POST',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted({ name }, { dispatch, queryFulfilled }) {
        if (!name) return; // Guard against null/undefined name
        const patchResult = dispatch(
          cityApiSlice.util.updateQueryData('getCities', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const cities = draft as any;
            const tempId = `temp-${Date.now()}`;
            cities.push({ _id: tempId, name });
            cities.sort((a: City, b: City) => a.name.localeCompare(b.name));
          })
        );

        try {
          const { data: newCity } = await queryFulfilled;
          // Replace temporary city with server response
          dispatch(
            cityApiSlice.util.updateQueryData('getCities', undefined, (draft) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              const cities = draft as any;
              const index = cities.findIndex((c: City) => c._id.startsWith('temp-'));
              if (index !== -1) {
                cities[index] = newCity;
                cities.sort((a: City, b: City) => a.name.localeCompare(b.name));
              }
            })
          );
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['City'] as any,
    }),
    updateCity: builder.mutation<City, UpdateCityData>({
      query: (credentials) => ({
        url: '/city',
        method: 'PATCH',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted({ _id, name }, { dispatch, queryFulfilled }) {
        if (!_id || !name) return; // Guard against null/undefined values
        const patchResult = dispatch(
          cityApiSlice.util.updateQueryData('getCities', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const cities = draft as any;
            const city = cities.find((c: City) => c._id === _id);
            if (city) {
              city.name = name;
              cities.sort((a: City, b: City) => a.name.localeCompare(b.name));
            }
          })
        );

        try {
          const { data: updatedCity } = await queryFulfilled;
          // Update with server response to ensure canonical data
          dispatch(
            cityApiSlice.util.updateQueryData('getCities', undefined, (draft) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              const cities = draft as any;
              const index = cities.findIndex((c: City) => c._id === _id);
              if (index !== -1) {
                cities[index] = updatedCity;
                cities.sort((a: City, b: City) => a.name.localeCompare(b.name));
              }
            })
          );
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['City'] as any,
    }),
    deleteCity: builder.mutation<{ success: boolean }, { _id: string }>({
      query: (credentials) => ({
        url: '/city',
        method: 'DELETE',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted({ _id }, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          cityApiSlice.util.updateQueryData('getCities', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const cities = draft as any;
            const index = cities.findIndex((c: City) => c._id === _id);
            if (index !== -1) {
              cities.splice(index, 1);
            }
          })
        );

        try {
          await queryFulfilled;
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['City'] as any,
    }),
    uploadImage: builder.mutation<{ url: string }, { file: File }>({
      query: credentials => {
        const formData = new FormData();
        console.log('credentials', credentials);
        formData.append('avatar', credentials.file);
        return {
          url: '/upload',
          method: 'POST',
          body: formData,
        };
      },
    }),
  }),
  overrideExisting: true,
});

export const {
  useGetCitiesQuery,
  useAddCityMutation,
  useUpdateCityMutation,
  useDeleteCityMutation,
  useUploadImageMutation,
} = cityApiSlice;