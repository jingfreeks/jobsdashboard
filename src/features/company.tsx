import { apiSlice } from '@/config/apiSplice';

// Type definitions
export interface Company {
  _id: string;
  name: string;
  address?: string;
  cityId?: string;
  cityname?: string;
}

export interface CompanyFormData {
  name: string;
  address?: string;
  cityId?: string;
}

export interface UpdateCompanyData extends CompanyFormData {
  _id: string;
}

export const companyApiSlice = apiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getCompanies: builder.query<Company[], void>({
      query: () => '/company',
      transformResponse: (response: Company[]) => {
        // Sort companies alphabetically by name
        return response.sort((a, b) => a.name.localeCompare(b.name));
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['Company'] as any,
      keepUnusedDataFor: 300, // Cache for 5 minutes
    }),
    addCompany: builder.mutation<Company, CompanyFormData>({
      query: (credentials) => ({
        url: '/company',
        method: 'POST',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted(credentials, { dispatch, queryFulfilled }) {
        if (!credentials || !credentials.name) return; // Guard against null/undefined credentials
        const { name } = credentials;
        const patchResult = dispatch(
          companyApiSlice.util.updateQueryData('getCompanies', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const companies = draft as any;
            const tempId = `temp-${Date.now()}`;
            companies.push({ _id: tempId, name });
            companies.sort((a: Company, b: Company) => a.name.localeCompare(b.name));
          })
        );

        try {
          const { data: newCompany } = await queryFulfilled;
          // Replace temporary company with server response
          dispatch(
            companyApiSlice.util.updateQueryData('getCompanies', undefined, (draft) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              const companies = draft as any;
              const index = companies.findIndex((c: Company) => c._id.startsWith('temp-'));
              if (index !== -1) {
                companies[index] = newCompany;
                companies.sort((a: Company, b: Company) => a.name.localeCompare(b.name));
              }
            })
          );
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Company'] as any,
    }),
    updateCompany: builder.mutation<Company, UpdateCompanyData>({
      query: (credentials) => ({
        url: '/company',
        method: 'PATCH',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted(credentials, { dispatch, queryFulfilled }) {
        if (!credentials || !credentials._id || !credentials.name) return; // Guard against null/undefined values
        const { _id, name } = credentials;
        const patchResult = dispatch(
          companyApiSlice.util.updateQueryData('getCompanies', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const companies = draft as any;
            const company = companies.find((c: Company) => c._id === _id);
            if (company) {
              company.name = name;
              companies.sort((a: Company, b: Company) => a.name.localeCompare(b.name));
            }
          })
        );

        try {
          const { data: updatedCompany } = await queryFulfilled;
          // Update with server response to ensure canonical data
          dispatch(
            companyApiSlice.util.updateQueryData('getCompanies', undefined, (draft) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              const companies = draft as any;
              const index = companies.findIndex((c: Company) => c._id === _id);
              if (index !== -1) {
                companies[index] = updatedCompany;
                companies.sort((a: Company, b: Company) => a.name.localeCompare(b.name));
              }
            })
          );
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Company'] as any,
    }),
    deleteCompany: builder.mutation<{ success: boolean }, { _id: string }>({
      query: (credentials) => ({
        url: '/company',
        method: 'DELETE',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted({ _id }, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          companyApiSlice.util.updateQueryData('getCompanies', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const companies = draft as any;
            const index = companies.findIndex((c: Company) => c._id === _id);
            if (index !== -1) {
              companies.splice(index, 1);
            }
          })
        );

        try {
          await queryFulfilled;
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Company'] as any,
    }),
  }),
  overrideExisting: true,
});

export const {
  useGetCompaniesQuery,
  useAddCompanyMutation,
  useUpdateCompanyMutation,
  useDeleteCompanyMutation,
} = companyApiSlice;

// Performance-optimized selectors
export const selectCompaniesSorted = (data: Company[] | undefined) => {
  if (!data) return [];
  return [...data].sort((a, b) => a.name.localeCompare(b.name));
};

export const selectCompanyById = (data: Company[] | undefined, id: string) => {
  if (!data) return undefined;
  return data.find(company => company._id === id);
};