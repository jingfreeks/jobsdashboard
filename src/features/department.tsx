import { apiSlice } from '@/config/apiSplice';

// Types
export interface Department {
  _id: string;
  name: string;
}

export interface DepartmentFormData {
  name: string;
}

export interface UpdateDepartmentData extends DepartmentFormData {
  _id: string;
}

// API slice with performance optimizations and optimistic updates
export const departmentApiSlice = apiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getDepartments: builder.query<Department[], void>({
      query: () => '/dept',
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['Department'] as any,
      // Cache for 5 minutes
      keepUnusedDataFor: 300,
      // Transform and sort data for better performance
      transformResponse: (response: Department[]) => {
        return response.sort((a, b) => a.name.localeCompare(b.name));
      },
    }),
    addDepartment: builder.mutation<Department, DepartmentFormData>({
      query: (credentials) => ({
        url: '/dept',
        method: 'POST',
        body: credentials,
      }),
      // Optimistic update for better UX
      async onQueryStarted(credentials, { dispatch, queryFulfilled }) {
        const tempId = `temp-${Date.now()}`;
        const optimisticDepartment: Department = {
          _id: tempId,
          name: credentials.name || '',
        };

        const patchResult = dispatch(
          departmentApiSlice.util.updateQueryData('getDepartments', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const departments = draft as any;
            departments.push(optimisticDepartment);
            departments.sort((a: Department, b: Department) => a.name.localeCompare(b.name));
          })
        );

        try {
          const { data: newDepartment } = await queryFulfilled;
          // Update with real data from server
          dispatch(
            departmentApiSlice.util.updateQueryData('getDepartments', undefined, (draft) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              const departments = draft as any;
              const index = departments.findIndex((dept: Department) => dept._id === tempId);
              if (index !== -1) {
                departments[index] = newDepartment;
              }
            })
          );
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Department'] as any,
    }),
    updateDepartment: builder.mutation<Department, UpdateDepartmentData>({
      query: (credentials) => ({
        url: '/dept',
        method: 'PATCH',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted({ _id, name }, { dispatch, queryFulfilled }) {
        if (!_id || !name) return; // Guard against null/undefined values
        const patchResult = dispatch(
          departmentApiSlice.util.updateQueryData('getDepartments', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const departments = draft as any;
            const department = departments.find((d: Department) => d._id === _id);
            if (department) {
              department.name = name;
              departments.sort((a: Department, b: Department) => a.name.localeCompare(b.name));
            }
          })
        );

        try {
          const { data: updatedDepartment } = await queryFulfilled;
          // Update with server response to ensure canonical data
          dispatch(
            departmentApiSlice.util.updateQueryData('getDepartments', undefined, (draft) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              const departments = draft as any;
              const index = departments.findIndex((d: Department) => d._id === _id);
              if (index !== -1) {
                departments[index] = updatedDepartment;
                departments.sort((a: Department, b: Department) => a.name.localeCompare(b.name));
              }
            })
          );
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Department'] as any,
    }),
    deleteDepartment: builder.mutation<void, { _id: string }>({
      query: (credentials) => ({
        url: '/dept',
        method: 'DELETE',
        body: credentials,
      }),
      // Optimistic update
      async onQueryStarted({ _id }, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          departmentApiSlice.util.updateQueryData('getDepartments', undefined, (draft) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const departments = draft as any;
            const index = departments.findIndex((dept: Department) => dept._id === _id);
            if (index !== -1) {
              departments.splice(index, 1);
            }
          })
        );

        try {
          await queryFulfilled;
        } catch {
          // Revert optimistic update on error
          patchResult.undo();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Department'] as any,
    }),
  }),
  overrideExisting: true,
});

// Export hooks
export const {
  useGetDepartmentsQuery,
  useAddDepartmentMutation,
  useUpdateDepartmentMutation,
  useDeleteDepartmentMutation,
} = departmentApiSlice;

// Performance-optimized selectors
export const selectDepartmentsSorted = (data: Department[] | undefined) => {
  if (!data) return [];
  return [...data].sort((a, b) => a.name.localeCompare(b.name));
};

export const selectDepartmentById = (data: Department[] | undefined, id: string) => {
  if (!data) return undefined;
  return data.find(department => department._id === id);
};