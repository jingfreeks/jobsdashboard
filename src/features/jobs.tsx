import { apiSlice } from '@/config/apiSplice';

// Job interface
export interface Job {
  _id: string;
  jobtitle: string;
  companyId?: string;
  companyname?: string;
  compname?:string;
  cityId?: string;
  cityname?: string;
  departmentId?: string;
  departmentname?: string;
  shiftId?: string;
  shiftname?: string;
  description?: string;
  jobDescription?:string;
  jobRequirements?: string;
  requirements?: string;
  salary?: string;
  jobType?: string; // full-time, part-time, contract, etc.
  status?: string; // active, inactive, draft, etc.
}

// Form data interfaces
export interface JobFormData {
  jobtitle: string;
  companyId?: string;
  cityId?: string;
  departmentId?: string;
  description?: string;
  requirements?: string;
  salary?: string;
  type?: string;
  status?: string;
}

export interface UpdateJobData extends JobFormData {
  _id: string;
}

export const jobsApiSlice = apiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getJobs: builder.query<Job[], void>({
      query: () => '/jobs',
      keepUnusedDataFor: 120,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['Jobs'] as any,
    }),

    getJobById: builder.query<Job, string>({
      query: (id) => `/jobs/${id}`,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: (_result, _error, id) => [{ type: 'Jobs', id }] as any,
    }),

    addJob: builder.mutation<Job, JobFormData>({
      query: (credentials) => ({
        url: '/jobs',
        method: 'POST',
        body: { ...credentials },
      }),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Jobs'] as any,
    }),

    updateJob: builder.mutation<Job, UpdateJobData>({
      query: (credentials) => ({
        url: `/jobs/${credentials._id}`,
        method: 'PATCH',
        body: { ...credentials },
      }),
       
      invalidatesTags: (_result, _error, { _id }) => [
        { type: 'Jobs', id: 'LIST' },
        { type: 'Jobs', id: _id },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ] as any,
    }),

    deleteJob: builder.mutation<{ success: boolean }, { _id: string }>({
      query: (credentials) => ({
        url: `/jobs`,
        method: 'DELETE',
        body: {...credentials },
      }),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      invalidatesTags: ['Jobs'] as any,
    }),

    getJobsByCompany: builder.query<Job[], string>({
      query: (companyId) => `/jobs/company/${companyId}`,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['CompanyJobs'] as any,
    }),

    getJobsByCity: builder.query<Job[], string>({
      query: (cityId) => `/jobs/city/${cityId}`,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['CityJobs'] as any,
    }),

    getJobsByDepartment: builder.query<Job[], string>({
      query: (departmentId) => `/jobs/department/${departmentId}`,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      providesTags: ['DepartmentJobs'] as any,
    }),
  }),
  overrideExisting: true,
});

export const {
  useGetJobsQuery,
  useGetJobByIdQuery,
  useAddJobMutation,
  useUpdateJobMutation,
  useDeleteJobMutation,
  useGetJobsByCompanyQuery,
  useGetJobsByCityQuery,
  useGetJobsByDepartmentQuery,
} = jobsApiSlice;