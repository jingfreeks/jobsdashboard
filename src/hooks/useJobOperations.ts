import { useMemo } from 'react';
import { 
  useGetJobsQuery, 
  useAddJobMutation, 
  useUpdateJobMutation, 
  useDeleteJobMutation,
  type Job,
  type JobFormData,
  type UpdateJobData
} from '@/features/jobs';
import { useGetCompaniesQuery } from '@/features/company';
import { useGetCitiesQuery } from '@/features/city';
import { useGetDepartmentsQuery } from '@/features/department';

export const useJobOperations = () => {
  // Fetch jobs data
  const { 
    data: jobs = [], 
    isLoading, 
    error, 
    refetch 
  } = useGetJobsQuery();

  // Fetch related data
  const { data: companies = [] } = useGetCompaniesQuery();
  const { data: cities = [] } = useGetCitiesQuery();
  const { data: departments = [] } = useGetDepartmentsQuery();

  // Mutations
  const [addJob, { isLoading: isAdding }] = useAddJobMutation();
  const [updateJob, { isLoading: isUpdating }] = useUpdateJobMutation();
  const [deleteJob, { isLoading: isDeleting }] = useDeleteJobMutation();

  // Create maps for efficient lookups
  const companyMap = useMemo(() => {
    return new Map(companies.map(company => [company._id, company]));
  }, [companies]);

  const cityMap = useMemo(() => {
    return new Map(cities.map(city => [city._id, city]));
  }, [cities]);

  const departmentMap = useMemo(() => {
    return new Map(departments.map(department => [department._id, department]));
  }, [departments]);

  console.log('Company Map:', jobs[0]);
  // Enhanced jobs with related data
  const jobsWithDetails = useMemo(() => {
    return jobs.map(job => ({
      ...job,
      companyname: job.companyId && companyMap.get(job.companyId)?.name || 'No Company',
      cityname: job.cityId && cityMap.get(job.cityId)?.name || 'No City',
      departmentname: job.departmentId && departmentMap.get(job.departmentId)?.name || 'No Department'
    }));
  }, [jobs, companyMap, cityMap, departmentMap]);

  // Job operations
  const createJob = async (jobData: JobFormData): Promise<Job | null> => {
    try {
      const result = await addJob(jobData).unwrap();
      return result;
    } catch (error) {
      console.error('Failed to create job:', error);
      return null;
    }
  };

  const updateJobById = async (jobData: UpdateJobData): Promise<Job | null> => {
    try {
      const result = await updateJob(jobData).unwrap();
      return result;
    } catch (error) {
      console.error('Failed to update job:', error);
      return null;
    }
  };

  const deleteJobById = async (jobId: string): Promise<boolean> => {
    try {
      const result = await deleteJob(jobId).unwrap();
      return result.success;
    } catch (error) {
      console.error('Failed to delete job:', error);
      return false;
    }
  };

  return {
    // Data
    jobs,
    jobsWithDetails,
    companies,
    cities,
    departments,
    companyMap,
    cityMap,
    departmentMap,
    
    // Loading states
    isLoading,
    isAdding,
    isUpdating,
    isDeleting,
    
    // Error state
    error,
    
    // Operations
    createJob,
    updateJobById,
    deleteJobById,
    refetch,
  };
}; 