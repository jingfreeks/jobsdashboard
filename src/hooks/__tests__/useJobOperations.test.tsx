/* eslint-disable @typescript-eslint/no-explicit-any */

// src/features/jobs/__tests__/useJobOperations.test.tsx
import { renderHook, act } from "@testing-library/react";
import { vi, describe, it, expect, beforeEach } from "vitest";
import { useJobOperations } from "../useJobOperations";
import type { Mock } from "vitest";
// Mock RTK Query hooks
vi.mock("@/features/jobs", () => ({
  useGetJobsQuery: vi.fn(),
  useAddJobMutation: vi.fn(),
  useUpdateJobMutation: vi.fn(),
  useDeleteJobMutation: vi.fn(),
}));

vi.mock("@/features/company", () => ({
  useGetCompaniesQuery: vi.fn(),
}));

vi.mock("@/features/city", () => ({
  useGetCitiesQuery: vi.fn(),
}));

vi.mock("@/features/department", () => ({
  useGetDepartmentsQuery: vi.fn(),
}));

// Import after mocks
import {
  useGetJobsQuery,
  useAddJobMutation,
  useUpdateJobMutation,
  useDeleteJobMutation,
} from "@/features/jobs";
import { useGetCompaniesQuery } from "@/features/company";
import { useGetCitiesQuery } from "@/features/city";
import { useGetDepartmentsQuery } from "@/features/department";

describe("useJobOperations (Vitest)", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("maps jobs with related company/city/department names", () => {
    (useGetJobsQuery as any).mockReturnValue({
      data: [
        { _id: "1", companyId: "c1", cityId: "ct1", departmentId: "d1" },
      ],
      isLoading: false,
      error: null,
      refetch: vi.fn(),
    });

    (useGetCompaniesQuery as Mock).mockReturnValue({
      data: [{ _id: "c1", name: "OpenAI" }],
    });
    (useGetCitiesQuery as Mock).mockReturnValue({
      data: [{ _id: "ct1", name: "San Francisco" }],
    });
    (useGetDepartmentsQuery as Mock).mockReturnValue({
      data: [{ _id: "d1", name: "Engineering" }],
    });

    (useAddJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);
    (useUpdateJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);
    (useDeleteJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);

    const { result } = renderHook(() => useJobOperations());

    expect(result.current.jobsWithDetails[0]).toMatchObject({
      companyname: "OpenAI",
      cityname: "San Francisco",
      departmentname: "Engineering",
    });

    // snapshot check
    expect(result.current.jobsWithDetails).toMatchSnapshot();
  });

  it("creates a job successfully", async () => {
    const mockAdd = vi.fn().mockReturnValue({
      unwrap: () => Promise.resolve({ _id: "1", title: "New Job" }),
    });
    (useAddJobMutation as Mock).mockReturnValue([mockAdd, { isLoading: false }]);

    (useGetJobsQuery as Mock).mockReturnValue({
      data: [],
      isLoading: false,
      error: null,
      refetch: vi.fn(),
    });
    (useGetCompaniesQuery as Mock).mockReturnValue({ data: [] });
    (useGetCitiesQuery as Mock).mockReturnValue({ data: [] });
    (useGetDepartmentsQuery as Mock).mockReturnValue({ data: [] });
    (useUpdateJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);
    (useDeleteJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);

    const { result } = renderHook(() => useJobOperations());

    let job;
    await act(async () => {
      job = await result.current.createJob({ title: "New Job" } as any);
    });

    expect(mockAdd).toHaveBeenCalled();
    expect(job).toEqual({ _id: "1", title: "New Job" });
  });

  it("handles deleteJob failure gracefully", async () => {
    const mockDelete = vi.fn().mockReturnValue({
      unwrap: () => Promise.reject(new Error("Delete failed")),
    });
    (useDeleteJobMutation as Mock).mockReturnValue([mockDelete, { isLoading: false }]);

    (useGetJobsQuery as Mock).mockReturnValue({
      data: [],
      isLoading: false,
      error: null,
      refetch: vi.fn(),
    });
    (useGetCompaniesQuery as Mock).mockReturnValue({ data: [] });
    (useGetCitiesQuery as Mock).mockReturnValue({ data: [] });
    (useGetDepartmentsQuery as Mock).mockReturnValue({ data: [] });
    (useAddJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);
    (useUpdateJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);

    const { result } = renderHook(() => useJobOperations());

    let success;
    await act(async () => {
      success = await result.current.deleteJobById("1");
    });

    expect(success).toBe(false);
  });

  it("updates a job successfully with updateJobById", async () => {
    const payload = { _id: "1", title: "Updated" } as any;
    const updated = { _id: "1", title: "Updated", status: "ok" };

    const mockUpdate = vi.fn().mockReturnValue({
      unwrap: () => Promise.resolve(updated),
    });
    (useUpdateJobMutation as Mock).mockReturnValue([mockUpdate, { isLoading: false }]);

    // minimal mocks for other deps
    (useGetJobsQuery as Mock).mockReturnValue({
      data: [],
      isLoading: false,
      error: null,
      refetch: vi.fn(),
    });
    (useGetCompaniesQuery as Mock).mockReturnValue({ data: [] });
    (useGetCitiesQuery as Mock).mockReturnValue({ data: [] });
    (useGetDepartmentsQuery as Mock).mockReturnValue({ data: [] });
    (useAddJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);
    (useDeleteJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);

    const { result } = renderHook(() => useJobOperations());

    let res;
    await act(async () => {
      res = await result.current.updateJobById(payload);
    });

    expect(mockUpdate).toHaveBeenCalledWith(payload);
    expect(res).toEqual(updated);
  });

  it("returns null when updateJobById fails", async () => {
    const payload = { _id: "1", title: "Broken" } as any;

    const mockUpdate = vi.fn().mockReturnValue({
      unwrap: () => Promise.reject(new Error("Update failed")),
    });
    (useUpdateJobMutation as Mock).mockReturnValue([mockUpdate, { isLoading: false }]);

    // minimal mocks for other deps
    (useGetJobsQuery as Mock).mockReturnValue({
      data: [],
      isLoading: false,
      error: null,
      refetch: vi.fn(),
    });
    (useGetCompaniesQuery as Mock).mockReturnValue({ data: [] });
    (useGetCitiesQuery as Mock).mockReturnValue({ data: [] });
    (useGetDepartmentsQuery as Mock).mockReturnValue({ data: [] });
    (useAddJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);
    (useDeleteJobMutation as Mock).mockReturnValue([vi.fn(), { isLoading: false }]);

    const { result } = renderHook(() => useJobOperations());

    let res;
    await act(async () => {
      res = await result.current.updateJobById(payload);
    });

    expect(mockUpdate).toHaveBeenCalledWith(payload);
    expect(res).toBeNull();
  });
});
