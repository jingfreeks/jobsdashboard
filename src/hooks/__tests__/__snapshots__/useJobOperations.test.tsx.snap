// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useJobOperations (Vitest) > maps jobs with related company/city/department names 1`] = `
[
  {
    "_id": "1",
    "cityId": "ct1",
    "cityname": "San Francisco",
    "companyId": "c1",
    "companyname": "OpenAI",
    "departmentId": "d1",
    "departmentname": "Engineering",
  },
]
`;
